---
title: "hw2"
author: "Nan Chen"
date: "2/2/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Q1. Data Analysis  
### 1. Exercise 7.3.4  
- Explore the distribution of each of the x, y, and z variables in diamonds. What do you learn? Think about a diamond and how you might decide which dimension is the length, width, and depth.
```{r}
if (!"tidyverse" %in% rownames(installed.packages()))
  install.packages("tidyverse", repos = "http://cra.rstudio.com/")
library(tidyverse)
ggplot(data = diamonds, mapping = aes(x = x)) +
  geom_histogram(binwidth = 0.1)

ggplot(data = diamonds, mapping = aes(x = y)) +
  geom_histogram(binwidth = 0.1) +
  coord_cartesian(xlim = c(0, 10))

ggplot(data = diamonds, mapping = aes(x = z)) +
  geom_histogram(binwidth = 0.1) +
  coord_cartesian(xlim = c(0, 8))
```  
**Comments: 1) After histograms of x, y, z have been generated, I find all these three distributions have several outliners and have the highest peak at left side, which indicates that they are all right skewed. 2) All three distributions have have a bimodality. 3) Also, according to x (~3.5-~9), y (~3.5-~9) and z (~2-~5) 's range, I think variable z is the depth, y is the width and x is the length.**  

- Explore the distribution of price. Do you discover anything unusual or surprising? (Hint: Carefully think about the binwidth and make sure you try a wide range of values.)  
```{r}
ggplot(data = diamonds, mapping = aes(x = price)) +
  geom_histogram(binwidth = 4.5) +
  coord_cartesian(ylim = c(0, 80))
```
**Comments: 1) The distribution of price is right skewed, and has many peaks. 2) One of the surprising thing I can notice is that there are no counts at certain price, such as $1500.**  

- How many diamonds are 0.99 carat? How many are 1 carat? What do you think is the cause of the difference?  
```{r}
library("tidyverse")
diamonds %>%
  filter(carat >= 0.99, carat <= 1) %>%
  count(carat)

diamonds %>%
   filter(carat >= 0.9, carat <= 1.1) %>%
   count(carat) %>%
   print(n = 20)
```
**Comments: There are only 23 diamonds which carat is 0.99, while there are 1558 diamonds which carat is 1. The reason more 1 carat diamonds sold than 0.99 carat diamonds is that some diamonds carat values are being “rounded up”. That is to say, the difference in value per carat is less expensive as general. This is can be confirmed by the second graph which lists the prices of diamionds from 0.9 carat to 1.1 carat.**  

- Compare and contrast coord_cartesian() vs xlim() or ylim() when zooming in on a histogram. What happens if you leave binwidth unset? What happens if you try and zoom so only half a bar shows?  
```{r}
ggplot(data = diamonds, mapping = aes(x = carat)) +
  geom_histogram() +
  coord_cartesian(ylim = c(0, 100))

ggplot(data = diamonds, mapping = aes(x = carat)) +
  geom_histogram() +
  ylim (0, 100)
```
**Comments: coord_catesian does not remove unseen data points, while xlim and ylim removes unseen data points from histogram. If we leave binwidth unset, the histogram bars are combined together and cannot reavel useful information for us. **  

### 2. Exercise 7.4.1  
- What happens to missing values in a histogram? What happens to missing values in a bar chart? Why is there a difference?  
```{r}
diamonds2 <- diamonds %>%
  mutate(y = ifelse(y < 3 | y > 20, NA, y))
ggplot(data = diamonds2) +
  geom_histogram(binwidth = 0.1, mapping = aes(x = y))

diamonds %>%
  mutate(color = if_else(runif(n()) < 0.1, NA_character_, 
                         as.character(color))) %>%
  ggplot() +
  geom_bar(mapping = aes(x = color))
```
**Comments: Missing values will be removed in a histogram. In geom_bar, missing value is treated as another category.**  

- What does na.rm = TRUE do in mean() and sum()?  
```{r}
diamonds2 <- diamonds %>%
  mutate(y = ifelse(y < 3 | y > 20, NA, y))
a <- mean(diamonds2$y, na.rm = FALSE)
b <- sum(diamonds2$y, na.rm = TRUE)
a
b
```
**Comments: Missing values will not be calculated in mean() and sum(), so the output of mean() and sum() can both have actual values other than just NA.**  

### 3. Exercise 7.5.1.1  
- Use what you’ve learned to improve the visualisation of the departure times of cancelled vs. non-cancelled flights.  
```{r}
if (!"nycflights13" %in% rownames(installed.packages()))
  install.packages("nycflights13", repos = "http://cra.rstudio.com/")
library("nycflights13")
flights %>% 
  mutate(
    canceled = is.na(dep_time),
    sched_hour = sched_dep_time %/% 100,
    sched_min = sched_dep_time %% 100,
    sched_dep_time = sched_hour + sched_min / 60
  ) %>% 
  ggplot() +
    geom_boxplot(mapping = aes(y = sched_dep_time, x = canceled))
```  
**Comments: As we can see from the graph, the departure time of canceled flights is larger than that of scheduled flights.**  

- What variable in the diamonds dataset is most important for predicting the price of a diamond? How is that variable correlated with cut? Why does the combination of those two relationships lead to lower quality diamonds being more expensive?  
```{r}
attach(diamonds)
plot(price, carat, main="Price VS Carat", 
  	xlab="Price ", ylab="Carat ", type="p")

plot(cut, carat, main="Cut VS Carat", 
  	xlab="Cut ", ylab="Carat ", type="p")
abline(lm(cut~carat), col="red")
```  
**Comments: I think Carat is most important for predicting the price of a diamond. According to the fit line, we can see that the price decreases when the cut is from fair to ideal. **  

- Install the ggstance package, and create a horizontal boxplot. How does this compare to using coord_flip()?  
```{r}
if (!"ggstance" %in% rownames(installed.packages()))
  install.packages("ggstance", repos = "http://cra.rstudio.com/")
library("ggstance")
diamonds %>%
  ggplot(aes(cut, carat)) +
  geom_boxplot() +
  coord_flip()

diamonds %>%
  ggplot(aes(carat, cut)) +
  geom_boxploth()
```  
**Comments: According to these two plots, they look the same. But in the aesthetics the x and y are flipped from the previous case.**  

- One problem with boxplots is that they were developed in an era of much smaller datasets and tend to display a prohibitively large number of “outlying values”. One approach to remedy this problem is the letter value plot. Install the lvplot package, and try using geom_lv() to display the distribution of price vs cut. What do you learn? How do you interpret the plots?  
```{r}
if (!"lvplot" %in% rownames(installed.packages()))
  install.packages("lvplot", repos = "http://cra.rstudio.com/")
library(lvplot)
ggplot(diamonds, aes(x = cut, y = price)) + 
  geom_lv()
```  
**Comments: The boxes of the letter-value plot correspond to many more quantiles. And I think they can be useful for larger datasets as larger datasets can give precise estimates of quantiles beyond the quartiles.**  

- Compare and contrast geom_violin() with a facetted geom_histogram(), or a coloured geom_freqpoly(). What are the pros and cons of each method?  
```{r}
ggplot(data = diamonds, mapping = aes(x = cut, y = price)) +
  geom_violin() +
  coord_flip()

ggplot(data = diamonds, mapping = aes(x = price)) +
  geom_histogram() +
  facet_wrap(~ cut, ncol = 1, scales = "free_y")

ggplot(data = diamonds, mapping = aes(x = price, y = ..density..)) + 
  geom_freqpoly(mapping = aes(colour = cut), binwidth = 500)
```  
**Comments: The  faceted geom_histogram and geom_violin have similar plots. It is easy for us to distinguish differences in the overall shape of the distributions (skewness, central values, variance, etc). However, it is difficult to look up which category has the highest density for a given price because we cannot compare y axis from the plots. The geom_freqpoly is better to visually see the trend between compared variables. For example, it is easy for us to tell which cut has the highest price. However, the weakness is the overlapping between the cut. It is hard to tell the relationship among different distributions.**  

- If you have a small dataset, it’s sometimes useful to use geom_jitter() to see the relationship between a continuous and categorical variable. The ggbeeswarm package provides a number of methods similar to geom_jitter(). List them and briefly describe what each one does.  
```{r}
if (!"ggbeeswarm" %in% rownames(installed.packages()))
  install.packages("ggbeeswarm", repos = "http://cra.rstudio.com/")
library(ggbeeswarm)
ggplot(data = mpg) +
  geom_quasirandom(mapping = aes(x = reorder(class, hwy, FUN = median), 
                                 y = hwy))
ggplot(data = mpg) +
  geom_quasirandom(mapping = aes(x = reorder(class, hwy, FUN = median), 
                                 y = hwy), method = "smiley")
ggplot(data = mpg) +
  geom_beeswarm(mapping = aes(x = reorder(class, hwy, FUN = median), y = hwy))
```  
**Comments: There are 2 methods. 1) geom_quasirandom that produces plots that resemble something between jitter and violin. There are several different methods that determine exactly how the random location of the points is generated. 2) geom_beeswarm creates a shape similar to a violin plot, but by offsetting the points.**  

### 4. Exercise 7.5.2.1  
- How could you rescale the count dataset above to more clearly show the distribution of cut within colour, or colour within cut?  
```{r}
if (!"viridis" %in% rownames(installed.packages()))
  install.packages("viridis", repos = "http://cra.rstudio.com/")
library("viridis")
diamonds %>%
  count(color, cut) %>%
  group_by(color) %>%
  mutate(scale = n / sum(n)) %>%
  ggplot(mapping = aes(x = color, y = cut)) +
  geom_tile(mapping = aes(fill = scale)) +
  scale_fill_viridis(limits = c(0, 1), option="plasma")

diamonds %>%
  count(color, cut) %>%
  group_by(cut) %>%
  mutate(scale = n / sum(n)) %>%
  ggplot(mapping = aes(x = cut, y = color)) +
  geom_tile(mapping = aes(fill = scale)) +
  scale_fill_viridis(limits = c(0, 1), option="plasma")
```  
**Comments: We need to first group by colour and calcualte the proportion scale of cut in each cut group, then we can see clearly the distribution of cut within colour. limits = c(0, 1) is added because it is easy for us to compare the differences among each group. Also, the same thing could be done to get the distribution of colour within cut.**  

- Use geom_tile() together with dplyr to explore how average flight delays vary by destination and month of year. What makes the plot difficult to read? How could you improve it?  
```{r}
flights %>%
  group_by(dest, month) %>%
  summarise(dep_delay = mean(dep_delay, na.rm = TRUE)) %>%
  ggplot(mapping = aes(x = factor(month), y = dest, fill = dep_delay)) +
  geom_tile() +
  labs(x = "Month", y = "Destination", fill = "Departure Delay")

#improved plots by viridis
flights %>%
  group_by(month, dest) %>%
  summarise(dep_delay = mean(dep_delay, na.rm = TRUE)) %>%
  group_by(dest) %>%
  filter(n() == 12) %>%
  ungroup() %>%
  mutate(dest = fct_reorder(dest, dep_delay)) %>%
  ggplot(aes(x = factor(month), y = dest, fill = dep_delay)) +
  geom_tile() +
  scale_fill_viridis(option = "plasma") +
  labs(x = "Month", y = "Destination", fill = "Departure Delay")
```  
**Comments: 1) We can sort destinations by average delay to make the plot easy for us to read. 2) We can remove missing values and use virids for a better colour panel.**  

- Why is it slightly better to use aes(x = color, y = cut) rather than aes(x = cut, y = color) in the example above?  
```{r}
diamonds %>% 
  count(color, cut) %>%  
  ggplot(mapping = aes(x = cut, y = color)) +
    geom_tile(mapping = aes(fill = n))
```  
**Comments: It is better to use the categorical variable with longer labels or larger number of categories on the y axis.**  

### 5. Exercise 7.5.3.1  
- Instead of summarising the conditional distribution with a boxplot, you could use a frequency polygon. What do you need to consider when using cut_width() vs cut_number()? How does that impact a visualisation of the 2d distribution of carat and price?  
```{r}
# cut_width()
ggplot(data = diamonds, 
       mapping = aes(x = price,
                     y = ..density.., 
                     colour = cut_width(carat, 0.3))) +
  geom_freqpoly()

# cut_number()
ggplot(data = diamonds, 
       mapping = aes(x = price,
                     y = ..density.., 
                     colour = cut_number(carat, 10))) +
  geom_freqpoly()
```  
**Comments: The number in each bin will be different if we use cut_width(). Since there are equal numbers in each bin, the plot looks the same if density is used for the y aesthetic.**  

- Visualize the distribution of carat, partitioned by price.  
```{r}
ggplot(diamonds, aes(x = cut_number(price, 10), y = carat)) +
  geom_boxplot() +
  coord_flip() +
  xlab("Price")
```  
**Comments: This box plot is partitioned into an 10 bins with the same number of observations.**

- How does the price distribution of very large diamonds compare to small diamonds. Is it as you expect, or does it surprise you?  
**Comments: The distribuyion of large diamonds varies a lot, but it does not surprise me because I think above a certain size other factors such as cut and colour will play more of a role in the price.**  

- Combine two of the techniques you’ve learned to visualize the combined distribution of cut, carat, and price.  
```{r}
ggplot(diamonds, mapping = aes(color = cut_number(carat, 5),x = cut, 
                               y = price)) +
  geom_boxplot()
```  

- Two dimensional plots reveal outliers that are not visible in one dimensional plots. For example, some points in the plot below have an unusual combination of x and y values, which makes the points outliers even though their x and y values appear normal when examined separately.  
```{r}
ggplot(data = diamonds) +
  geom_point(mapping = aes(x = x, y = y)) +
  coord_cartesian(xlim = c(3, 12), ylim = c(3, 12))
```  
**Comments: There is a strong relationship between x and y so the outliers are not extreme in either x or y. However, the bin plot will not reveal these outliers and mislead us that the largest x is the outlier.**


## Q3. Genetic Data Analysis  
### 1. How many persons are in the data set (statisticians call this n)? How many SNPs are in the data set (statisticians call this p)?  
```{r}

#read in data as tibble
merge.geno.bim <- read_tsv("/home/m280-data/hw1/merge-geno.bim", 
                           col_names = c("chromosome", "snpid", "gendist", "bp",
                                         "allelle1", "allelle2"))
merge.geno.fam <- read_delim("/home/m280-data/hw1/merge-geno.fam", delim = " ",
                             col_names = c("famid", "personid", "fid", "mid",
                                         "sex", "affstat"))

#count the number of persons in fam
merge.geno.fam %>%
  summarise(persons = n())
#count the number of SNPs in bim
merge.geno.bim %>%
  summarise(SNPs = n())
```  

### 2. Which chromosomes does this data set contain? How many SNPs are in each chromosome?  
```{r}
merge.geno.bim %>%
  group_by(chromosome) %>%
  summarise(SNPs = n())
```  

### 3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp – 48,130,769 bp. How many SNPs are located within MAP4 gene?  
```{r}
merge.geno.bim %>%
  filter(chromosome == 3 & (bp >= 47892180 & bp <= 48130769)) %>%
  summarise(SNPs = n())
```  

### 4. Convert File 
- Convert merge-geno.bim to Mendel SNP definition file
```{r}
merge.geno.bim1 <- merge.geno.bim %>%
  select(snpid, chromosome, bp)
lines <- c("    2.40 = FILE FORMAT VERSION NUMBER.", 
           "8348674 = NUMBER OF SNPS LISTED HERE.")
writeLines(lines, "dfile.txt", sep = "\n")
dfile1 <- write_delim(merge.geno.bim1, 'dfile.txt', delim = ",", na = "NA", 
                      append = TRUE, col_names = FALSE)
```
```{bash}
head dfile.txt
```

- Convert merge-geno.fam to Mendel pedigree file  
```{r}
merge.geno.fam1 <- mutate(merge.geno.fam, 
                      person_id_sub = substr(personid, 5, 13), 
                      father_id_sub = substr(fid, 5, 13),
                      mother_id_sub = substr(mid, 5, 13),
                      affection_status_sub = substr(affstat, 5, 13),
                      sex_sub = ifelse(sex == 1, "M", "F"))
pfile <- write.table(merge.geno.fam1[, c("famid", "person_id_sub", 
                                         "father_id_sub", "mother_id_sub", 
                                         "sex_sub", "affection_status_sub")],
              file = 'pfile.txt', sep = ",", na = "NA", row.names = FALSE, 
              quote = FALSE)
```
```{bash}
head pfile.txt
```














